File: ocaml/lib/rad_expr.ml
(** Radical expression AST, parameterised by the coefficient type.

    Normalization is explicit: these constructors are "dumb".
    Use {!Normalize} for simplification. *)

(** A radical expression over a coefficient field ['k]. *)
type 'k t =
  | Lit of 'k
  | Neg of 'k t
  | Add of 'k t * 'k t
  | Mul of 'k t * 'k t
  | Inv of 'k t
  | Root of int * 'k t    (** [Root (n, x)] = n-th root of x, n >= 2 *)
  | Pow of 'k t * int

(** {2 Comparison}

    Structural comparison — required for use as map/set keys.
    This is NOT mathematical equality (use {!Equality} for that). *)

let rec compare cmp_k a b =
  match a, b with
  | Lit x, Lit y -> cmp_k x y
  | Lit _, _ -> -1 | _, Lit _ -> 1
  | Neg x, Neg y -> compare cmp_k x y
  | Neg _, _ -> -1 | _, Neg _ -> 1
  | Add (a1, a2), Add (b1, b2) ->
    let c = compare cmp_k a1 b1 in
    if c <> 0 then c else compare cmp_k a2 b2
  | Add _, _ -> -1 | _, Add _ -> 1
  | Mul (a1, a2), Mul (b1, b2) ->
    let c = compare cmp_k a1 b1 in
    if c <> 0 then c else compare cmp_k a2 b2
  | Mul _, _ -> -1 | _, Mul _ -> 1
  | Inv x, Inv y -> compare cmp_k x y
  | Inv _, _ -> -1 | _, Inv _ -> 1
  | Root (n1, x), Root (n2, y) ->
    let c = Int.compare n1 n2 in
    if c <> 0 then c else compare cmp_k x y
  | Root _, _ -> -1 | _, Root _ -> 1
  | Pow (x, n1), Pow (y, n2) ->
    let c = compare cmp_k x y in
    if c <> 0 then c else Int.compare n1 n2

let rec equal eq_k a b =
  match a, b with
  | Lit x, Lit y -> eq_k x y
  | Neg x, Neg y -> equal eq_k x y
  | Add (a1, a2), Add (b1, b2) -> equal eq_k a1 b1 && equal eq_k a2 b2
  | Mul (a1, a2), Mul (b1, b2) -> equal eq_k a1 b1 && equal eq_k a2 b2
  | Inv x, Inv y -> equal eq_k x y
  | Root (n1, x), Root (n2, y) -> n1 = n2 && equal eq_k x y
  | Pow (x, n1), Pow (y, n2) -> equal eq_k x y && n1 = n2
  | _ -> false

(** {2 Construction helpers} *)

(** Wrap a coefficient as a literal. *)
let lit k = Lit k
(** Negate an expression. *)
let neg e = Neg e
(** Sum of two expressions. *)
let add a b = Add (a, b)
(** Subtraction: [sub a b = Add (a, Neg b)]. *)
let sub a b = Add (a, Neg b)
(** Product of two expressions. *)
let mul a b = Mul (a, b)
(** Division: [div' a b = Mul (a, Inv b)]. *)
let div' a b = Mul (a, Inv b)
(** Multiplicative inverse. *)
let inv e = Inv e
(** Principal [n]-th root of [e] (n >= 2). *)
let root n e = Root (n, e)
(** Square root shorthand. *)
let sqrt' e = Root (2, e)
(** Integer power. *)
let pow e n = Pow (e, n)

(** Lift an integer into a radical expression over rationals. *)
let of_int n = Lit (Rational.of_int n)
(** Lift a rational into a radical expression. *)
let of_rational r = Lit r

(** {2 Functor map (fmap)} *)

(** Map a function over all coefficients in the expression. *)
let rec map f = function
  | Lit k -> Lit (f k)
  | Neg a -> Neg (map f a)
  | Add (a, b) -> Add (map f a, map f b)
  | Mul (a, b) -> Mul (map f a, map f b)
  | Inv a -> Inv (map f a)
  | Root (n, a) -> Root (n, map f a)
  | Pow (a, n) -> Pow (map f a, n)

(** {2 Structural queries} *)

(** Nesting depth of the expression tree.  Literals have depth 0. *)
let rec depth = function
  | Lit _ -> 0
  | Neg a -> depth a
  | Add (a, b) -> 1 + max (depth a) (depth b)
  | Mul (a, b) -> 1 + max (depth a) (depth b)
  | Inv a -> 1 + depth a
  | Root (_, a) -> 1 + depth a
  | Pow (a, _) -> 1 + depth a

(** Number of nodes in the expression tree. *)
let rec size = function
  | Lit _ -> 1
  | Neg a -> 1 + size a
  | Add (a, b) -> 1 + size a + size b
  | Mul (a, b) -> 1 + size a + size b
  | Inv a -> 1 + size a
  | Root (_, a) -> 1 + size a
  | Pow (a, _) -> 1 + size a

(** Check if all coefficients satisfy a predicate. *)
let rec for_all_coeffs p = function
  | Lit k -> p k
  | Neg a -> for_all_coeffs p a
  | Add (a, b) -> for_all_coeffs p a && for_all_coeffs p b
  | Mul (a, b) -> for_all_coeffs p a && for_all_coeffs p b
  | Inv a -> for_all_coeffs p a
  | Root (_, a) -> for_all_coeffs p a
  | Pow (a, _) -> for_all_coeffs p a

(** Collect distinct [(root_index, radicand)] pairs. *)
let collect_radicals eq_k expr =
  let mem pair acc =
    List.exists (fun (n2, e2) ->
      let (n1, e1) = pair in
      n1 = n2 && equal eq_k e1 e2) acc
  in
  let rec go acc = function
    | Lit _ -> acc
    | Neg a -> go acc a
    | Add (a, b) -> go (go acc a) b
    | Mul (a, b) -> go (go acc a) b
    | Inv a -> go acc a
    | Pow (a, _) -> go acc a
    | Root (n, a) ->
      let acc' = go acc a in
      let pair = (n, a) in
      if mem pair acc' then acc' else acc' @ [pair]
  in
  go [] expr

File: mercury/rad_expr.m
%---------------------------------------------------------------------------%
% vim: ft=mercury ts=4 sw=4 et
%---------------------------------------------------------------------------%
% File: rad_expr.m
% Main author: nyc
%
% Core AST for radical expressions, parameterised by the coefficient type.
% Includes construction helpers and structural queries.
%
% Normalization is explicit: these constructors are "dumb".
% Call functions in rad_normalize for simplification.
%
%---------------------------------------------------------------------------%

:- module rad_expr.
:- interface.

:- import_module list.
:- import_module rational.

%---------------------------------------------------------------------------%
% Core type
%---------------------------------------------------------------------------%

    % A radical expression over a coefficient field K.
    %
:- type rad_expr(K)
    --->    re_lit(K)
    ;       re_neg(rad_expr(K))
    ;       re_add(rad_expr(K), rad_expr(K))
    ;       re_mul(rad_expr(K), rad_expr(K))
    ;       re_inv(rad_expr(K))
    ;       re_root(int, rad_expr(K))   % Root n x: principal nth root
    ;       re_pow(rad_expr(K), int).   % Pow x n: integer power

%---------------------------------------------------------------------------%
% Construction helpers
%---------------------------------------------------------------------------%

    % Subtraction: re_sub(A, B) = re_add(A, re_neg(B)).
    %
:- func re_sub(rad_expr(K), rad_expr(K)) = rad_expr(K).

    % Division: re_div(A, B) = re_mul(A, re_inv(B)).
    %
:- func re_div(rad_expr(K), rad_expr(K)) = rad_expr(K).

    % Square root shorthand: re_sqrt(X) = re_root(2, X).
    %
:- func re_sqrt(rad_expr(K)) = rad_expr(K).

    % Lift a rational number into a radical expression.
    %
:- func rat_e(rational) = rad_expr(rational).

    % Lift a machine integer into a radical expression over rationals.
    %
:- func int_e(int) = rad_expr(rational).

%---------------------------------------------------------------------------%
% Structural queries
%---------------------------------------------------------------------------%

    % Nesting depth of the expression tree.
    %
:- func re_depth(rad_expr(K)) = int.

    % Number of nodes in the expression tree.
    %
:- func re_size(rad_expr(K)) = int.

    % Check if all coefficients satisfy a predicate.
    %
:- pred re_free_of(pred(K), rad_expr(K)).
:- mode re_free_of(in(pred(in) is semidet), in) is semidet.

    % Map a function over the coefficients.
    %
:- func re_map_coeffs((func(K1) = K2), rad_expr(K1)) = rad_expr(K2).

%---------------------------------------------------------------------------%
% Radical collection and dependency ordering
%---------------------------------------------------------------------------%

    % Collect distinct (rootIndex, radicand) pairs from an expression.
    %
:- func collect_radicals(rad_expr(K)) = list({int, rad_expr(K)}).

    % Topologically sort radicals so that radicals with rational radicands
    % come first, followed by radicals whose radicands depend only on
    % earlier radicals.
    %
:- func topo_sort_radicals(list({int, rad_expr(K)}))
    = list({int, rad_expr(K)}).

    % Check whether all Root subexpressions in a radicand are present
    % in the resolved set.
    %
:- pred all_roots_resolved(list({int, rad_expr(K)})::in,
    rad_expr(K)::in) is semidet.

%---------------------------------------------------------------------------%
%---------------------------------------------------------------------------%

:- implementation.

:- use_module int.

%---------------------------------------------------------------------------%
% Construction helpers
%---------------------------------------------------------------------------%

re_sub(A, B) = re_add(A, re_neg(B)).
re_div(A, B) = re_mul(A, re_inv(B)).
re_sqrt(A) = re_root(2, A).
rat_e(R) = re_lit(R).
int_e(N) = re_lit(rational(N, 1)).

%---------------------------------------------------------------------------%
% Structural queries
%---------------------------------------------------------------------------%

re_depth(re_lit(_)) = 0.
re_depth(re_neg(A)) = re_depth(A).
re_depth(re_add(A, B)) = int.'+'(1, int.max(re_depth(A), re_depth(B))).
re_depth(re_mul(A, B)) = int.'+'(1, int.max(re_depth(A), re_depth(B))).
re_depth(re_inv(A)) = int.'+'(1, re_depth(A)).
re_depth(re_root(_, A)) = int.'+'(1, re_depth(A)).
re_depth(re_pow(A, _)) = int.'+'(1, re_depth(A)).

re_size(re_lit(_)) = 1.
re_size(re_neg(A)) = int.'+'(1, re_size(A)).
re_size(re_add(A, B)) = int.'+'(1, int.'+'(re_size(A), re_size(B))).
re_size(re_mul(A, B)) = int.'+'(1, int.'+'(re_size(A), re_size(B))).
re_size(re_inv(A)) = int.'+'(1, re_size(A)).
re_size(re_root(_, A)) = int.'+'(1, re_size(A)).
re_size(re_pow(A, _)) = int.'+'(1, re_size(A)).

re_free_of(P, re_lit(K)) :- P(K).
re_free_of(P, re_neg(A)) :- re_free_of(P, A).
re_free_of(P, re_add(A, B)) :- re_free_of(P, A), re_free_of(P, B).
re_free_of(P, re_mul(A, B)) :- re_free_of(P, A), re_free_of(P, B).
re_free_of(P, re_inv(A)) :- re_free_of(P, A).
re_free_of(P, re_root(_, A)) :- re_free_of(P, A).
re_free_of(P, re_pow(A, _)) :- re_free_of(P, A).

re_map_coeffs(F, Expr) = Result :-
    (
        Expr = re_lit(K),
        Result = re_lit(F(K))
    ;
        Expr = re_neg(A),
        Result = re_neg(re_map_coeffs(F, A))
    ;
        Expr = re_add(A, B),
        Result = re_add(re_map_coeffs(F, A), re_map_coeffs(F, B))
    ;
        Expr = re_mul(A, B),
        Result = re_mul(re_map_coeffs(F, A), re_map_coeffs(F, B))
    ;
        Expr = re_inv(A),
        Result = re_inv(re_map_coeffs(F, A))
    ;
        Expr = re_root(N, A),
        Result = re_root(N, re_map_coeffs(F, A))
    ;
        Expr = re_pow(A, N),
        Result = re_pow(re_map_coeffs(F, A), N)
    ).

%---------------------------------------------------------------------------%
% Radical collection and dependency ordering
%---------------------------------------------------------------------------%

collect_radicals(Expr) = nub_pairs(go(Expr)) :-
    true.

:- func go(rad_expr(K)) = list({int, rad_expr(K)}).

go(re_lit(_)) = [].
go(re_neg(A)) = go(A).
go(re_add(A, B)) = go(A) ++ go(B).
go(re_mul(A, B)) = go(A) ++ go(B).
go(re_inv(A)) = go(A).
go(re_pow(A, _)) = go(A).
go(re_root(N, A)) = go(A) ++ [{N, A}].

:- func nub_pairs(list({int, rad_expr(K)})) = list({int, rad_expr(K)}).

nub_pairs([]) = [].
nub_pairs([X | Xs]) = [X | nub_pairs(list.filter(
    ( pred(Y::in) is semidet :- not X = Y ), Xs))].

topo_sort_radicals(Radicals) = topo_loop([], Radicals).

:- func topo_loop(list({int, rad_expr(K)}),
    list({int, rad_expr(K)})) = list({int, rad_expr(K)}).

topo_loop(Sorted, []) = Sorted.
topo_loop(Sorted, Remaining) = Result :-
    Remaining = [_ | _],
    list.filter(( pred(R::in) is semidet :-
        {_, Rad} = R,
        all_roots_resolved(Sorted, Rad)
    ), Remaining, Ready, NotReady),
    ( if Ready = [] then
        Result = Sorted ++ Remaining
    else
        Result = topo_loop(Sorted ++ Ready, NotReady)
    ).

all_roots_resolved(_, re_lit(_)).
all_roots_resolved(Resolved, re_neg(A)) :-
    all_roots_resolved(Resolved, A).
all_roots_resolved(Resolved, re_add(A, B)) :-
    all_roots_resolved(Resolved, A),
    all_roots_resolved(Resolved, B).
all_roots_resolved(Resolved, re_mul(A, B)) :-
    all_roots_resolved(Resolved, A),
    all_roots_resolved(Resolved, B).
all_roots_resolved(Resolved, re_inv(A)) :-
    all_roots_resolved(Resolved, A).
all_roots_resolved(Resolved, re_pow(A, _)) :-
    all_roots_resolved(Resolved, A).
all_roots_resolved(Resolved, re_root(N, A)) :-
    list.member({N, A}, Resolved).

%---------------------------------------------------------------------------%
:- end_module rad_expr.
%---------------------------------------------------------------------------%
